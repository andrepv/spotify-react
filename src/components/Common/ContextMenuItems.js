import React, { Component } from "react";
import PropTypes from "prop-types";
import { ReactComponent as Loader } from "images/loader.svg";

export default class ContextMenuItems extends Component {
  constructor(props) {
    super(props);
    this.state = {
      searchInputValue: "",
      menuItems: props.items,
    };
  }

  componentDidUpdate(prevProps, prevState) {
    const {items, disableInfiniteScroll} = this.props;
    const {searchInputValue} = this.state;
    if (prevProps.items.length < items.length) {
      this.setState({
        menuItems: items,
      });
    }
    if (searchInputValue !== prevState.searchInputValue) {
      disableInfiniteScroll &&
      disableInfiniteScroll(Boolean(searchInputValue.length));
    }
  }

  resetState = () => {
    this.setState({
      searchInputValue: "",
      menuItems: this.props.items,
    });
  }

  selectItem = item => {
    if (item.name === "no results") {
     return;
    }
    this.props.handler({
      ...item,
      name: Array.isArray(item.name) ? item.name.join("") : item.name,
    });
    this.resetState();
  }

  handleInputChange = event => {
    const inputValue = event.target.value.toLowerCase().trim();
    let searchMatches = this.props.items;

    if (inputValue.length) {
      searchMatches = this.props.items.filter(item => {
        const itemName = item.name.toLowerCase();
        return itemName.includes(inputValue);
      }).map(item => {
        const reg = new RegExp(`(${inputValue})`, "i");
        return {
          ...item,
          name: item.name.split(reg),
        };
      });
    }

    this.setState({
      searchInputValue: event.target.value,
      menuItems: searchMatches.length ? searchMatches : [{name: "no results"}],
    });
  }

  renderItem(itemName) {
    if (Array.isArray(itemName)) {
      return itemName.map((part, index) => {
        const inputValue = this.state.searchInputValue.toLowerCase();
        const itemPart = part.toLowerCase();
        return (
          <span
            key={index}
            className={`${itemPart === inputValue ? "marked" : ""}`}
          >
            {part}
          </span>
        );
      });
    }
    return itemName;
  }

  renderMenuItems() {
    return this.state.menuItems.map((item, index) => {
      return (
        <li
          key={index}
          className="select__option overflow-ellipsis"
          onClick={() => this.selectItem(item)}
        >
          {this.renderItem(item.name)}
        </li>
      );
    });
  }

  render() {
    const {loadMorePending, items, emptyMsg} = this.props;
    if (!items.length) {
      return (
        <p className="select__empty">
          {emptyMsg}
        </p>
      );
    }
    return (
      <div className="select__dropdown">
        <div className="select__inner custom-scrollbar">
          <ul>
            {this.renderMenuItems()}
          </ul>
          {loadMorePending && <div className="loader"><Loader/></div>}
        </div>
        <div className="select__search">
          <input
            type="text"
            className="select__search-input"
            placeholder="Search..."
            onChange={this.handleInputChange}
            value={this.state.searchInputValue}
          />
        </div>
      </div>
    );
  }
}

ContextMenuItems.propTypes = {
  items: PropTypes.array.isRequired,
  handler: PropTypes.func.isRequired,
  disableInfiniteScroll: PropTypes.func,
  emptyMsg: PropTypes.string,
  loadMorePending: PropTypes.bool,
};