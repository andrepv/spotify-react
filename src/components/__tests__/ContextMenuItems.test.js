import React from "react";
import { shallow } from "enzyme";
import renderer from "react-test-renderer";
import ContextMenuItems from "components/Common/ContextMenuItems";

describe("<ContextMenuItems />", () => {
  it("renders correctly", () => {
    const props = {
      items: [{name: "item1"}, {name: "item2"}],
      handler: jest.fn(),
    };
    const component = renderer.create(
      <ContextMenuItems {...props} />
    ).toJSON();
    expect(component).toMatchSnapshot();
  });

  it("should render items correctly", () => {
    const props = {
      items: [{name: "item1"}, {name: "item2"}],
      handler: jest.fn(),
    };
    const wrapper = shallow(<ContextMenuItems {...props} />);
    const target = wrapper.find(".select__option");

    expect(wrapper.state("menuItems")).toEqual(props.items);
    expect(target).toHaveLength(props.items.length);
    expect(target.at(0).text()).toEqual(props.items[0].name);
  });

  it("should render the right search results", () => {
    const props = {
      items: [{name: "item1"}, {name: "item2"}],
      handler: jest.fn(),
      disableInfiniteScroll: jest.fn(),
    };
    const wrapper = shallow(<ContextMenuItems {...props} />);
    wrapper.find(".select__search-input").simulate(
      "change",
      {target: {value: "i"}}
    );

    const items = wrapper.find(".select__option");
    const searchInput = wrapper.find(".select__search-input");
    expect(items).toHaveLength(props.items.length);
    expect(items.everyWhere(n => {
      return (
        n.children().length === 3 &&
        n.find("span.marked").text() === "i" &&
        n.find("span:last-child").text().includes("tem")
      );
    })).toEqual(true);

    expect(props.disableInfiniteScroll.mock.calls.length).toBe(1);
    expect(props.disableInfiniteScroll).lastCalledWith(true);

    items.at(0).simulate("click");
    expect(props.handler.mock.calls.length).toBe(1);

    searchInput.at(0).simulate(
      "change",
      {target: {value: ""}}
    );
    expect(wrapper.state("menuItems")).toEqual(props.items);
    expect(items.at(0).text()).toEqual(props.items[0].name);
    items.at(0).simulate("click");
    expect(props.handler.mock.calls.length).toBe(2);
  });

  it("should render a message when no search results", () => {
    const props = {
      items: [{name: "item1"}, {name: "item2"}],
      handler: jest.fn(),
    };
    const wrapper = shallow(<ContextMenuItems {...props} />);
    wrapper.find(".select__search-input").at(0).simulate(
      "change",
      {target: {value: "o"}}
    );
    const items = wrapper.find(".select__option");

    expect(items).toHaveLength(1);
    expect(items.at(0).text()).toEqual("no results");

    items.at(0).simulate("click");
    expect(props.handler.mock.calls.length).toBe(0);
  });

  it("should render a message when no items", () => {
    const props = {
      items: [],
      handler: jest.fn(),
      emptyMsg: "empty",
    };
    const wrapper = shallow(<ContextMenuItems {...props} />);
    const target = wrapper.find(".select__empty");
    expect(target).toHaveLength(1);
    expect(target.at(0).text()).toEqual(props.emptyMsg);
  });

  it("should render a loader", () => {
    const props = {
      items: [{name: "item1"}],
      handler: jest.fn(),
      loadMorePending: true,
    };
    const wrapper = shallow(<ContextMenuItems {...props} />);
    expect(wrapper.find(".loader")).toHaveLength(1);
  });
});