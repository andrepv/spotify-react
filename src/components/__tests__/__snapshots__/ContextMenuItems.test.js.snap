// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ContextMenuItems /> renders correctly 1`] = `
<div
  className="select__dropdown"
>
  <div
    className="select__inner custom-scrollbar"
  >
    <ul>
      <li
        className="select__option overflow-ellipsis"
        onClick={[Function]}
      >
        item1
      </li>
      <li
        className="select__option overflow-ellipsis"
        onClick={[Function]}
      >
        item2
      </li>
    </ul>
  </div>
  <div
    className="select__search"
  >
    <input
      className="select__search-input"
      onChange={[Function]}
      placeholder="Search..."
      type="text"
      value=""
    />
  </div>
</div>
`;
